### PR TITLE
[GCS] refactor gcs initialization

### PR DESCRIPTION
**Pull request by wumuzi520**
## Why are these changes needed?
- Problem
After the GCS server is restarted, multiple modules need to recover data from the storage. If there is a dependency between modules, the following callback hell will appear.
```
auto load_completed_count = std::make_shared<int>(0);
  int load_count = 2;
  auto on_done = [this, load_count, load_completed_count]() {
    ++(*load_completed_count);

    // We will reschedule the unfinished actors, so we have to load the actor data at the
    // end to make sure the other table data is loaded.
    if (*load_completed_count == load_count) {
      auto actor_manager_load_initial_data_callback = [this]() {
           // ...
      };
      gcs_actor_manager_->LoadInitialData(actor_manager_load_initial_data_callback);
    }
  };
  gcs_object_manager_->LoadInitialData(on_done);
  gcs_node_manager_->LoadInitialData(on_done);
```
As the module complexity increases, there may be three or more callback layers.

In addition, if the state of all GCS modules has not been restored, we should not make the dependencies on each other take effect, otherwise it will complicate the initialization process. 

- Solution
  - Loads all required metadata from the store into memory at once, so that the next
initialization process can be synchronized.
  - Install listeners after all all modules are initialized.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
@wumuzi520 Please remove the [WIP] once it is ready (I added). Also, please update the description.

**Comment by wumuzi520** [comment]
> @wumuzi520 Please remove the [WIP] once it is ready (I added). Also, please update the description.

I think it is ready to review now, thanks :)

File: src/ray/gcs/gcs_server/gcs_init_data.cc [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_init_data.h"
+
+namespace ray {
+namespace gcs {
+void GcsInitData::AsyncLoad(const EmptyCallback &on_done) {
+  // There are 5 kinds of table data need to be loaded.
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_init_data.cc:20]: // There are 5 kinds of table data need to be loaded.
```suggestion
  // There are 6 kinds of table data need to be loaded.
```

File: src/ray/gcs/gcs_server/gcs_init_data.h [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  const std::unordered_map<NodeID, rpc::ResourceMap> &ClusterResources() const {
+    return resource_table_data_;
+  }
+
+  /// Get actor metadata.
+  const std::unordered_map<ActorID, rpc::ActorTableData> &Actors() const {
+    return actor_table_data_;
+  }
+
+  const std::unordered_map<PlacementGroupID, rpc::PlacementGroupTableData>
```
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_init_data.h:67]: /// Get placement group metadata.

**Review by ffbin — APPROVED** [review]
LGTM, just some small comments. BTW: Pls merge master and fix the testcase of ci job.

**Review by raulchen — APPROVED** [review]
LGTM. much cleaner

**Pull request merged**